crates/ruff_server/src/trace.rs

```diff
@@ -0,0 +1,143 @@
+}
+
+pub(crate) fn init_tracing(
+    sender: ClientSender,
+    log_level: LogLevel,
+    log_file: Option<&std::path::Path>,
+) {
+    LOGGING_SENDER
+        .set(sender)
+        .expect("logging sender should only be initialized once");
+
+    let subscriber = tracing_subscriber::Registry::default().with(
+        tracing_subscriber::fmt::layer()
+            .with_timer(Uptime::default())
+            .with_thread_names(true)
+            .with_ansi(false)
+            .with_writer(TracingWriter::new(log_file))
+            .with_filter(TraceLevelFilter)
+            .with_filter(LogLevelFilter { filter: log_level }),
+    );
+
+    tracing::subscriber::set_global_default(subscriber)
+        .expect("should be able to set global default subscriber");
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(rename_all = "lowercase")]
+pub(crate) enum LogLevel {
+    #[default]
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+impl LogLevel {
+    fn trace_level(self) -> tracing::Level {
+        match self {
+            Self::Error => tracing::Level::ERROR,
+            Self::Warn => tracing::Level::WARN,
+            Self::Info => tracing::Level::INFO,
+            Self::Debug => tracing::Level::DEBUG,
+            Self::Trace => tracing::Level::TRACE,
+        }
+    }
+}
+
+struct LogLevelFilter {
+    filter: LogLevel,
+}
+
+struct TraceLevelFilter;
+
+impl<S> tracing_subscriber::layer::Filter<S> for LogLevelFilter {
+    fn enabled(
+        &self,
+        meta: &tracing::Metadata<'_>,
+        _: &tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        let filter = if meta.target().starts_with("ruff") {
+            self.filter.trace_level()
+        } else {
+            tracing::Level::INFO
+        };
+
+        meta.level() <= &filter
+    }
+
+    fn max_level_hint(&self) -> Option<tracing::level_filters::LevelFilter> {
+        Some(LevelFilter::from_level(self.filter.trace_level()))
+    }
+}
+
+impl<S> tracing_subscriber::layer::Filter<S> for TraceLevelFilter {
+    fn enabled(
+        &self,
+        _: &tracing::Metadata<'_>,
+        _: &tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        trace_value() != lsp_types::TraceValue::Off
+    }
+}
+
+struct TracingWriter(Option<PathBuf>);
+
+impl TracingWriter {
+    fn new(file: Option<&std::path::Path>) -> TracingWriter {
+        Self(file.map(std::path::Path::to_path_buf))
+    }
+}
+
+impl tracing_subscriber::fmt::MakeWriter<'_> for TracingWriter {
+    type Writer = Box<dyn std::io::Write>;
+
+    fn make_writer(&self) -> Self::Writer {
+        if let Some(file) = self.0.as_ref().and_then(|path| {
+            std::fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(path)
+                .ok()
+        }) {
+            Box::new(file)
+        } else {
+            Box::new(std::io::stderr())
+        }
+    }
+}
```

Comment:
Nit: `tracing-subscriber` has a `BoxMakeWriter` that is just that. You can call it with `BoxMakeWriter::new(std::io::stderr)`

Reply:
Unfortunately, this is a different return type, so it wouldn't work here. We need to consolidate the log file / stderr writer into a single type here, otherwise the tracing layer would have a different type and we'd need to build it in two completely distinct branches.

Reply:
It seems to work fine for rust analyzer:

```
    let writer = match log_file {
        Some(file) => BoxMakeWriter::new(Arc::new(file)),
        None => BoxMakeWriter::new(std::io::stderr),
    };
```

Reply:
Oh, I see what you mean! Nevermind 😄